gapis/resolve/dependencygraph2/dependency_graph.go

```diff
@@ -0,0 +1,426 @@
+	initMemoryNodeIDs obsNodeIDs
+	initCmdObsNodeIDs []obsNodeIDs
+	cmdObsNodeIDs     []obsNodeIDs
+	initialCommands   []api.Cmd
+	nodes             []Node
+	numDependencies   uint64
+	dependenciesFrom  [][]NodeID
+	dependenciesTo    [][]NodeID
+
+	config DependencyGraphConfig
+}
+
+// newDependencyGraph constructs a new dependency graph
+func newDependencyGraph(ctx context.Context, config DependencyGraphConfig,
+	c *capture.Capture, initialCmds []api.Cmd) *dependencyGraph {
+	g := &dependencyGraph{
+		capture:         c,
+		cmdNodeIDs:      new(api.SubCmdIdxTrie),
+		initialCommands: initialCmds,
+		config:          config,
+	}
+	numCmds := len(initialCmds) + len(c.Commands)
+	numObservations := countObservations(initialCmds) + countObservations(c.Commands)
+	numNodes := numCmds + numObservations
+	g.nodes = make([]Node, 0, numNodes)
+	g.initCmdObsNodeIDs = make([]obsNodeIDs, len(initialCmds))
+	for i, cmd := range initialCmds {
+		g.initCmdObsNodeIDs[i] = g.addCmdNodes(api.CmdID(i).Derived(), cmd.Extras().Observations())
+	}
+	g.cmdObsNodeIDs = make([]obsNodeIDs, len(c.Commands))
+	for i, cmd := range c.Commands {
+		g.cmdObsNodeIDs[i] = g.addCmdNodes(api.CmdID(i), cmd.Extras().Observations())
+	}
+	g.dependenciesFrom = make([][]NodeID, numNodes)
+	return g
+}
+
+// NumNodes returns the number of nodes in the graph
+func (g *dependencyGraph) NumNodes() int {
+	return len(g.nodes)
+}
+
+// NumDependencies returns the number of dependencies (edges) in the graph
+func (g *dependencyGraph) NumDependencies() uint64 {
+	return g.numDependencies
+}
+
+// GetNode returns the node data associated with the given NodeID
+func (g *dependencyGraph) GetNode(nodeID NodeID) Node {
+	if nodeID >= NodeID(len(g.nodes)) {
+		return nil
+	}
+	return g.nodes[nodeID]
+}
+
+// GetNodeID returns the NodeID associated with given node data
+func (g *dependencyGraph) GetNodeID(node Node) NodeID {
+	if cmdNode, ok := node.(CmdNode); ok {
+		index := cmdNode.Index
+		if len(index) == 0 {
+			return NodeNoID
+		}
+		if !g.config.IncludeInitialCommands && !api.CmdID(index[0]).IsReal() {
+			return NodeNoID
+		}
+		if g.config.MergeSubCmdNodes {
+			index = index[:1]
+		}
+		if val := g.cmdNodeIDs.Value(index); val != nil {
+			return val.(NodeID)
+		}
+	} else if obsNode, ok := node.(ObsNode); ok {
+		cmdID := obsNode.CmdID
+		var obsNodeIDs obsNodeIDs
+		if cmdID == api.CmdNoID {
+			obsNodeIDs = g.initMemoryNodeIDs
+		} else if cmdID.IsReal() {
+			obsNodeIDs = g.cmdObsNodeIDs[cmdID]
+		} else {
+			cmdID = cmdID.Real()
+			obsNodeIDs = g.initCmdObsNodeIDs[cmdID]
+		}
+		index := obsNode.Index
+		if obsNode.IsWrite {
+			return obsNodeIDs.writeNodeIDStart + NodeID(index)
+		} else {
+			return obsNodeIDs.readNodeIDStart + NodeID(index)
+		}
+	}
+	return NodeNoID
+}
+
+// ForeachCmd iterates over all API calls in the graph
+func (g *dependencyGraph) ForeachCmd(ctx context.Context, cb func(context.Context, api.CmdID, api.Cmd) error) error {
```

Comment:
Why can't we use the `ForeachCmd` in `gapis/api`?

Reply:
This version loops over both the initial commands as well as the captured commands, and assigns a derived CmdID to the initial commands.

I can refactor the implementation of dependencyGraph.ForeachCmd to use api.ForeachCmd, rather than duplicating that logic here.

I'll also make the comment more explicit.